### PR TITLE
Create a test Ample implementation

### LINKED ISSUE #4376: Need a way to test UNKNOWN result for conditional mutations
Conditional mutations are used for metadata updates now.  Conditional mutations can return a result of [UNKNOWN](https://github.com/apache/accumulo/blob/9550da7c838994f9f6e9547819566b01e4b1dc6e/core/src/main/java/org/apache/accumulo/core/client/ConditionalWriter.java#L114).  The Ample code that handles conditional updates has [a method](https://github.com/apache/accumulo/blob/9550da7c838994f9f6e9547819566b01e4b1dc6e/core/src/main/java/org/apache/accumulo/core/metadata/schema/Ample.java#L631) that accepts a lambda for handling these unknown conditions.  This [code](https://github.com/apache/accumulo/blob/9550da7c838994f9f6e9547819566b01e4b1dc6e/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java#L1346) shows an example lambda.

These lambdas in the code currently have no test.  Need to devise a way to test these and then start writing test for all of the conditions.  It is possible that Ample may need to change to make this condition more testable.

### PR DESCRIPTION
**Pull request by cshannon**
Marking this as a draft for now as there is still work to be done and has various TODOs. I plan to continue working on this over the next week but wanted to push up what I have. Even after this gets merged I still expect there to be follow on issues to enhance the test framework as we need more features.

This creates a test Ample implementation that can be used to interact with Ample against a different table than the standard metadata table that the Accumulo instance uses. This will allow testing Ample without having to deal with metadata being written to the table by the system.

The test Ample also will support injecting custom behavior such as changing the return status of a conditional mutation in order to test. Right now the test implementation only supports the USER DataLevel to avoid having to deal with ZK. This may stay like this as the algorithms for testing should be the same for any DataLevel.

There is an example Test IT that shows how to use the API to test, including an example showing how to test with UNKNOWN status when using conditional mutations.

This closes #4376

### CODE REVIEW

File: test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+      } catch (Exception e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    /**
+     * Create metadata for a table by copying existing metadata for the table from the metadata
+     * table in an existing Accumulo instance
+     *
+     * TODO: Add config parents (such as a way to include/exclude what is copied, etc)
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java:123]: Could pass a `BiPredicate<Key, Value>` to the function.  Not sure if I have ever used BiPredicate, this seems like the perfect place for it.

File: test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                      e.getValue());
+                });
+            bw.addMutation(m);
+          }
+        }
+      }
+    }
+
+    public ConditionalTabletsMutator
+        conditionallyMutateTablets(ConditionalWriterInterceptor interceptor) {
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java:162]: Having this interceptor will be great for testing.  Need some way to use it when code only has a handle to the Ample interface.   Wonder if we could pass something like `Supplier< ConditionalWriterInterceptor> interceptorFactory` to the  TestAmple constructor.   Then could do the following in TestAmple

```java
    @Override
    public ConditionalTabletsMutator conditionallyMutateTablets() {
      return conditionallyMutateTablets(interceptorFactory.get());
    }
```

**Comment by cshannon** [comment]
The latest update [demonstrates](https://github.com/apache/accumulo/blob/e0fe9e49b9bc00f101afa6a7fa302f792d53b55b/test/src/main/java/org/apache/accumulo/test/ample/TestAmpleIT.java#L203) using the new test API that the fate unspittable marker is written for an unsplittlable tablet. It essentially replicates parts of the test in [LargeSplitRowIT](https://github.com/apache/accumulo/blob/eea8ce48e2d5d7bd5c6b69e05c6824f697d8de0a/test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java#L262) that runs with the real metadata table. 

One really nice side effect of this that I didn't think about is that you can use break points. Normally when testing with ITs the FATE code is running in the Manager JVM so you would need to open up a remote debug port to use break points. With this change the test repo code is running in the unit test JVM so you can easily step through with break points the code which makes debugging a lot easier.

I think we need both types of tests going forward but certainly having this new API will be nice to test things that are not easy with a normal IT.

File: test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java (hunk 1 above)
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java:168]: I like this idea, I just pushed up a new commit that implements it https://github.com/apache/accumulo/pull/4415/commits/e6b199eed5cf2ae1dc20a2d7873a2e49843bb964

File: test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java (hunk 2 above)
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java:213]: Added in [e5feb49](https://github.com/apache/accumulo/pull/4415/commits/e5feb4960ccc06430330650ec057dcd8aab5adad)

**Comment by cshannon** [comment]
The latest [commit](https://github.com/apache/accumulo/pull/4415/commits/6180ec183e54c543b00e101a5f496a10bf7b6640) includes a test for #3908, this could be broken out but thought it was a nice demonstration of the API to add it as part of this PR

File: test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  public static Ample create(ServerContext context, Map<DataLevel,String> tables) {
+    return new TestServerAmpleImpl(context, tables);
+  }
+
+  public static class TestServerAmpleImpl extends ServerAmpleImpl {
+
+    private final Map<DataLevel,String> tables;
+
+    public TestServerAmpleImpl(ServerContext context, final Map<DataLevel,String> tables) {
+      super(context);
```
[the remark this answers is not in the document]
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/ample/TestAmple.java]: I added support for using a tableMapper function in the latest commit 064a96452455672f8d5d82ad67beb95cb4929ac3
It did seem to make things a bit simpler in TestAmple now that the table mappings can be provided

**Comment by cshannon** [comment]
I am marking this as ready for review as I think it's pretty close to being ready to merge after all the recent changes. Because it's a test API I expect it to continue to mutate and evolve as we use it for more test cases but I think it's a good start in its current state. I added a couple tests for #3908 as stated in another comment.

File: server/base/src/test/java/org/apache/accumulo/server/metadata/TestAmple.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                getContext().createConditionalWriter(getTableMapper().apply(dataLevel)),
+                interceptor);
+          }
+        }
+      };
+    }
+
+    @Override
+    ServerContext getContext() {
+      // TODO: we can probably cache this
```
Comment (keith-turner) [inline server/base/src/test/java/org/apache/accumulo/server/metadata/TestAmple.java:213]: Could Suppliers.memoize be used for this?

**Review by keith-turner — APPROVED** [review]

**Pull request merged**